### PR TITLE
Ehcache base class to mutualize basic and loader-writer code

### PR DESCRIPTION
**Pull request by henri-tremblay**
This is the first step to mutualize the code between Ehcache and EhcacheWithLoaderWriter. A lot more can be done but at least we now have a starting point. This mutualization will be highly useful for implementing resilience.

### CODE REVIEW

File: impl/src/test/java/org/ehcache/docs/Ehcache3.java [hunk 1]
```diff
@@ -41,7 +39,7 @@ public void ehcache3Expiry() throws Exception {
     CacheConfigurationBuilder<Long, String> configuration =
         CacheConfigurationBuilder.newCacheConfigurationBuilder(Long.class, String.class, ResourcePoolsBuilder
             .heap(100))
-            .withExpiry(new Expiry<Long, String>() {    // <1>
+            .withExpiry(new ExpiryPolicy<Long, String>() {    // <1>
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/docs/Ehcache3.java]: Good catch!

File: core/src/main/java/org/ehcache/core/Ehcache.java [hunk 2]
```diff
@@ -21,38 +21,26 @@
 import java.util.ArrayList;
 import java.util.Collection;
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/Ehcache.java]: Imports above this one are unused

File: core/src/main/java/org/ehcache/core/EhcacheBase.java [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Map<BulkOps, LongAdder> getBulkMethodEntries() {
+    return bulkMethodEntries;
+  }
+
+  protected static void checkNonNull(Object thing) {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheBase.java]: Can't we leverage Java 8 constructs for these?

File: core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java [hunk 4]
```diff
@@ -1019,79 +831,9 @@ public boolean replace(final K key, final V oldValue, final V newValue) throws C
...
-  }
-
-  /**
-   * {@inheritDoc}
-   */
   @Override
-  public Jsr107Cache<K, V> getJsr107Cache() {
-    return jsr107Cache;
+  public Jsr107Cache<K, V> createJsr107Cache() {
+    return new Jsr107CacheImpl();
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java]: Why this change? There now can be multiple instances of that type for a given Cache. And I am not sure that does not break some (untested) spec wording. So unless you have a strong argument for this, I would tend to leave it as it was - that is a field initialised during construction.

File: api/src/main/java/org/ehcache/config/CacheConfiguration.java [hunk 5]
```diff
@@ -92,7 +92,7 @@
    * @deprecated Use {@link #getExpiryPolicy()}
    */
   @Deprecated
-  Expiry<? super K, ? super V> getExpiry();
+  org.ehcache.expiry.Expiry<? super K, ? super V> getExpiry();
```
[the remark this answers is not in the document]
Comment (henri-tremblay) [reply api/src/main/java/org/ehcache/config/CacheConfiguration.java]: This is a separate commit. It fixes deprecation warnings caused by #2166. Here is need to use the full path because I can't suppress warnings on imports. A stupid Java quirks.

File: core/src/main/java/org/ehcache/core/Ehcache.java (hunk 2 above)
Comment (henri-tremblay) [reply core/src/main/java/org/ehcache/core/Ehcache.java]: Indeed, merge side effect. Thanks

File: core/src/main/java/org/ehcache/core/EhcacheBase.java (hunk 3 above)
Comment (henri-tremblay) [reply core/src/main/java/org/ehcache/core/EhcacheBase.java]: Yes

File: core/src/main/java/org/ehcache/core/EhcacheRuntimeConfiguration.java [hunk 6]
```diff
@@ -100,8 +99,9 @@ public ClassLoader getClassLoader() {
     return this.classLoader;
   }
 
+  @SuppressWarnings("deprecation")
   @Override
-  public Expiry<? super K, ? super V> getExpiry() {
+  public org.ehcache.expiry.Expiry<? super K, ? super V> getExpiry() {
```
[the remark this answers is not in the document]
Comment (henri-tremblay) [reply core/src/main/java/org/ehcache/core/EhcacheRuntimeConfiguration.java]: Exactly

File: core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java (hunk 4 above)
Comment (henri-tremblay) [reply core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java]: I can't. Because I can't this do `super(new Jsr107CacheImpl(this))`. The alternative is to set `this` using a setter. Which is ugly. But no, it doesn't break anything because the JSR wrapper will call `createJsr107Cache` once. It's even better because it is only initialize when JSR107 is used.

File: api/src/main/java/org/ehcache/config/CacheConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ljacomet) [reply api/src/main/java/org/ehcache/config/CacheConfiguration.java]: But then where is the suppress warnings for that instance?

File: core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java (hunk 4 above)
Comment (ljacomet) [reply core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java]: Or have a field in subclasses for this ...

Comment (henri-tremblay) [reply core/src/main/java/org/ehcache/core/EhcacheWithLoaderWriter.java]: Yes. That would work too. But I preferred this way. Isn't it more efficient to instantiate it when needed?

File: api/src/main/java/org/ehcache/config/CacheConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (henri-tremblay) [reply api/src/main/java/org/ehcache/config/CacheConfiguration.java]: It seems that since the method itself is deprecated, there's no need to suppress the warning
